### ISSUE TITLE
render API

### ISSUE THREAD

**Issue by Fil** [observablehq/plot#501]
https://github.com/observablehq/plot/pull/498 has introduced function marks: `() => svg node`

That render function is called like any other and receives as input (data, scales, channels, dimensions), with empty data and channels, all the instantiated scales, and complete dimensions.

It would be useful I think to design and document this API (for all marks). The changes I'd suggest would be:

1. pass the facet's "content" (i.e. "Chinstrap")
2. order the API as (dimensions, scales, facet, data, channels)
3. change the dimensions in such a way that we can ignore margins (ie width would be the current width-marginLeft-marginRight, and the <g> would already be translated by marginLeft, marginTop)

https://observablehq.com/d/cfdce446a293784f

**Comment by mbostock** [comment]
We can’t change this API without breaking backwards compatibility, so I do not wish to redesign it. I agree we should document _mark_.render though.

**Comment by Fil** [comment]
Would you be against passing *facet* as a fifth argument?

**Comment by mbostock** [comment]
I realized we are inconsistent with what we pass mark.render. Without faceting, we pass {index, scales, values, dimensions, axes}; with faceting we pass {index, scales, values, dimensions}. So, if we want to pass the current facet it would need to be a sixth argument, or we’d somehow need to squirrel it into one of the existing arguments. But aside from how I’m onboard with somehow exposing the current facet values to marks during rendering.

**Comment by Fil** [comment]
One of the recurring frustrations with the current API is that _render_ doesn't know who the parent SVG and the parent FIGURE will be.

This makes it awkward to build interactivity, since we have to wait (asynchronously) for the `g` we return to be appended to the plot before we can bind listeners, or know where to dispatch our custom events to update the viewof.

It's still probably too early to cast this in concrete, but it should be included in the discussion about the API. Having the 6th argument be {facet} would allow to extend it in the future to {facet, parentSVG, figure} or something else (a callback when the figure is added to the DOM, maybe).

**Comment by Fil** [comment]
I did a bit of documentation in https://observablehq.com/d/cfdce446a293784f ; but def. not a blocker for 0.2

**Comment by Fil** [comment]
superseded by #1263
